### PR TITLE
IGNITE-15240 Thin client: Implement transactions

### PR DESCRIPTION
**Pull request by ptupitsyn**
* Add TX_BEGIN, TX_COMMIT, TX_ROLLBACK operations to the protocol.
* Add `ClientResourceRegistry` to store active transactions.
* Pass transaction id from all table API variants.

### CODE REVIEW

**Review by isapego — CHANGES_REQUESTED** [review]
As `ClientResourceRegistry` is connection-local, we should bind every transaction to a specific connection on client side and route all requests using this transaction through this connection. Also, let's add at least one more node to tests to make sure our implementation works in multi-node cluster :)

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/tx/ClientTransactionCommitRequest.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+     * Processes the request.
+     *
+     * @param in        Unpacker.
+     * @param resources Resources.
+     * @return Future.
+     */
+    public static CompletableFuture<Void> process(ClientMessageUnpacker in, ClientResourceRegistry resources) {
+        long resourceId = in.unpackLong();
+
+        Transaction t = resources.remove(resourceId).get();
```
[the remark this answers is not in the document]
Comment (ptupitsyn) [reply modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/tx/ClientTransactionCommitRequest.java]: 1. There is no NPE, `ClientResourceRegistry.remove` will throw a proper error message.
2. Added type check to `ClientResource`.

**Comment by ptupitsyn** [comment]
@isapego 

Good catch regarding connection/transaction ownership! Somehow forgot about it, even though we have `PayloadOutputChannel#clientChannel` exactly for this.

1. Added channel check to `ClientTable#writeTx` (same approach as we have in 2.x).
2. Enabled multi-node connection in `ItThinClientTransactionsTest`.

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/ClientResource.java [hunk 2]
```diff
@@ -17,20 +17,45 @@
...
     /**
      * Gets the underlying object.
      *
      * @param <T> Object type.
      * @return Object.
      */
-    <T> T get();
+    public <T> T get(Class<T> type) {
+        if (!type.isInstance(obj)) {
+            throw new IgniteException("Incorrect resource type. Expected " + type.getName() + ", but got " + obj.getClass().getName());
```
Comment (isapego) [inline modules/client-handler/src/main/java/org/apache/ignite/client/handler/ClientResource.java:47]: Why can not we get type information from `T`?

**Review by isapego — APPROVED** [review]
Overall, looks good to me

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/ClientResource.java (hunk 2 above)
Comment (ptupitsyn) [reply modules/client-handler/src/main/java/org/apache/ignite/client/handler/ClientResource.java:47]: Due to Java type erasure, `T` is replaced with `Object` at compile time, so there is no type information at run time.

**Pull request merged**